Signal/src/view controllers/SignalAttachment.swift

```diff
@@ -0,0 +1,463 @@
+                let imageData = UIPasteboard.general.data(forPasteboardType:dataUTI)
+                return videoAttachment(withData : imageData, dataUTI : dataUTI)
+            }
+        }
+        for dataUTI in audioUTISet() {
+            if pasteboardUTISet.contains(dataUTI) {
+                let imageData = UIPasteboard.general.data(forPasteboardType:dataUTI)
+                return audioAttachment(withData : imageData, dataUTI : dataUTI)
+            }
+        }
+        // TODO: We could handle generic attachments at this point.
+
+        return nil
+    }
+
+    // MARK: Image Attachments
+
+    // Factory method for an image attachment.
+    //
+    // NOTE: The attachment returned by this method may not be valid.
+    //       Check the attachment's error property.
+    public class func imageAttachment(withData imageData: Data?, dataUTI: String!) -> SignalAttachment! {
+        assert(dataUTI.characters.count > 0)
+
+        assert(imageData != nil)
+        guard let imageData = imageData else {
+            return nil
+        }
+
+        let attachment = SignalAttachment(data : imageData, dataUTI: dataUTI)
+
+        guard inputImageUTISet().contains(dataUTI) else {
+            attachment.error = .invalidFileFormat
+            return attachment
+        }
+
+        guard imageData.count > 0 else {
+            assert(imageData.count > 0)
+            attachment.error = .invalidData
+            return attachment
+        }
+
+        if dataUTI == kUTTypeGIF as String {
+            guard imageData.count <= kMaxFileSizeGif else {
+                attachment.error = .fileSizeTooLarge
+                return attachment
+            }
+            // We don't re-encode GIFs as JPEGs, presumably in case they are
+            // animated.
+            //
+            // TODO: Consider re-encoding non-animated GIFs as JPEG?
+            Logger.verbose("\(TAG) Sending raw \(attachment.mimeType()) to retain any animation")
+            return attachment
+        } else {
+            guard let image = UIImage(data:imageData) else {
+                attachment.error = .couldNotParseImage
+                return attachment
+            }
+            attachment.image = image
+
+            if isInputImageValidOutputImage(image: image, imageData: imageData, dataUTI: dataUTI) {
+                Logger.verbose("\(TAG) Sending raw \(attachment.mimeType())")
+                return attachment
+            }
+
+            Logger.verbose("\(TAG) Compressing attachment as image/jpeg")
+            return compressImageAsJPEG(image : image, attachment : attachment)
+        }
+    }
+
+    // If the proposed attachment already is a JPEG, and already conforms to the 
+    // file size and content size limits, don't recompress it.
+    //
+    // TODO: Should non-JPEGs always be converted to JPEG?
+    private class func isInputImageValidOutputImage(image: UIImage?, imageData: Data?, dataUTI: String!) -> Bool {
```

Comment:
One of the important fixes in this refactor is that we're now much more careful to avoid converting/re-encoding images as JPEG if possible. JPEG is a lossy format and needlessly re-encoding significantly degrades image quality.